services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/serialization/ResumableFileUploadSerializer.java

```diff
@@ -0,0 +1,180 @@
+        jsonGenerator.writeFieldName(UPLOAD_FILE_REQUEST);
+        jsonGenerator.writeStartObject();
+        jsonGenerator.writeFieldName(SOURCE);
+        jsonGenerator.writeValue(fileRequest.source().toString());
+        marshallPutObjectRequest(fileRequest.putObjectRequest(), jsonGenerator);
+        jsonGenerator.writeEndObject();
+    }
+
+    private static void marshallPutObjectRequest(PutObjectRequest putObjectRequest, JsonWriter jsonGenerator) {
+        jsonGenerator.writeFieldName(PUT_OBJECT_REQUEST);
+        jsonGenerator.writeStartObject();
+        validateNoRequestOverrideConfiguration(putObjectRequest);
+        putObjectRequest.sdkFields().forEach(field -> marshallPojoField(field, putObjectRequest, jsonGenerator));
+        jsonGenerator.writeEndObject();
+    }
+
+    private static void validateNoRequestOverrideConfiguration(PutObjectRequest putObjectRequest) {
+        if (putObjectRequest.overrideConfiguration().isPresent()) {
+            log.debug(() -> "ResumableFileUpload PutObjectRequest contains an override configuration that will not be "
+                            + "serialized");
+        }
+    }
+
+    private static void marshallPojoField(SdkField<?> field, PutObjectRequest request, JsonWriter jsonGenerator) {
+        Object val = field.getValueOrDefault(request);
+        TransferManagerJsonMarshaller<Object> marshaller = getMarshaller(field.marshallingType(), val);
+        marshaller.marshall(val, jsonGenerator, field.locationName());
+    }
+
+    public static ResumableFileUpload fromJson(String bytes) {
+        JsonNodeParser jsonNodeParser = JsonNodeParser.builder().build();
+        Map<String, JsonNode> uploadNodes = jsonNodeParser.parse(bytes).asObject();
+        return fromNodes(uploadNodes);
+    }
+
+    public static ResumableFileUpload fromJson(byte[] bytes) {
+        JsonNodeParser jsonNodeParser = JsonNodeParser.builder().build();
+        Map<String, JsonNode> uploadNodes = jsonNodeParser.parse(bytes).asObject();
+        return fromNodes(uploadNodes);
+    }
+
+    public static ResumableFileUpload fromJson(InputStream bytes) {
+        JsonNodeParser jsonNodeParser = JsonNodeParser.builder().build();
+        Map<String, JsonNode> uploadNodes = jsonNodeParser.parse(bytes).asObject();
+        return fromNodes(uploadNodes);
+    }
+
+    private static ResumableFileUpload fromNodes(Map<String, JsonNode> uploadNodes) {
+        TransferManagerJsonUnmarshaller<Object> longUnmarshaller = getUnmarshaller(MarshallingType.LONG);
+        TransferManagerJsonUnmarshaller<Object> instantUnmarshaller = getUnmarshaller(MarshallingType.INSTANT);
+        TransferManagerJsonUnmarshaller<Object> stringUnmarshaller = getUnmarshaller(MarshallingType.STRING);
+
+        ResumableFileUpload.Builder builder = ResumableFileUpload.builder();
+        builder.fileLength((Long) longUnmarshaller.unmarshall(uploadNodes.get(FILE_LENGTH)));
```

Comment:
No null check for `uploadNodes.get(FILE_LENGTH)`, what happens if its null ?

Reply:
fileLength should always be present. I can add null check here